### PR TITLE
Backport `Address` type to 0.3.x branch.

### PR DESCRIPTION
**Pull request by nuttycom**

### CODE REVIEW

**Review by ebfull — APPROVED** [review]

File: src/lib.rs [hunk 1]
```diff
@@ -203,6 +216,235 @@ impl TryFrom<u64> for Position {
...
+                index >>= 1;
+            } else {
+                return Address {
+                    level: Altitude(level),
+                    index,
+                };
+            }
+        }
+
+        unreachable!("The loop will always terminate via return in at most 64 iterations.")
```
Comment (daira) [inline src/lib.rs:386]:
```suggestion
        unreachable!("The loop will always terminate via return on the 65th iteration at the latest.")
```

File: src/lib.rs [hunk 2]
```diff
@@ -203,6 +216,235 @@ impl TryFrom<u64> for Position {
...
+impl From<Position> for Address {
+    fn from(p: Position) -> Self {
+        Address {
+            level: 0.into(),
+            index: p.into(),
+        }
+    }
+}
+
+impl<'a> From<&'a Position> for Address {
```
Comment (daira) [inline src/lib.rs:419]: Document that this converts to an Address at level 0.

File: src/lib.rs [hunk 3]
```diff
@@ -203,6 +216,235 @@ impl TryFrom<u64> for Position {
...
+    /// Increments this address's index by 1 and returns the resulting address.
+    pub fn next_at_level(&self) -> Address {
+        Address {
+            level: self.level,
+            index: self.index + 1,
+        }
+    }
+}
+
+impl From<Position> for Address {
```
Comment (daira) [inline src/lib.rs:410]: Document that this converts to an Address at level 0.

File: src/lib.rs [hunk 4]
```diff
@@ -203,6 +216,235 @@ impl TryFrom<u64> for Position {
...
+impl<'a> From<&'a Position> for Address {
+    fn from(p: &'a Position) -> Self {
+        Address {
+            level: 0.into(),
+            index: (*p).into(),
+        }
+    }
+}
+
+impl From<Address> for Option<Position> {
```
Comment (daira) [inline src/lib.rs:428]: Document.

File: src/lib.rs [hunk 5]
```diff
@@ -203,6 +216,235 @@ impl TryFrom<u64> for Position {
...
+    fn from(addr: Address) -> Self {
+        if addr.level == 0.into() {
+            Some(addr.index.into())
+        } else {
+            None
+        }
+    }
+}
+
+impl<'a> From<&'a Address> for Option<Position> {
```
Comment (daira) [inline src/lib.rs:438]: Document.

**Review by daira — APPROVED** [review]
utACK with suggestions.

**Comment by nuttycom** [comment]
Closed in favor of #62, which does the complete job.

**Pull request closed without merging**